src/core/server/saved_objects/routes/utils.ts

```diff
@@ -19,23 +19,30 @@
 
 import { Readable } from 'stream';
 import { SavedObject, SavedObjectsExportResultDetails } from 'src/core/server';
-import { createSplitStream, createMapStream, createFilterStream } from '../../utils/streams';
+import {
+  createSplitStream,
+  createMapStream,
+  createFilterStream,
+  createPromiseFromStreams,
+  createListStream,
+  createConcatStream,
+} from '../../utils/streams';
 
-export function createSavedObjectsStreamFromNdJson(ndJsonStream: Readable) {
-  return ndJsonStream
-    .pipe(createSplitStream('\n'))
-    .pipe(
-      createMapStream((str: string) => {
-        if (str && str.trim() !== '') {
-          return JSON.parse(str);
-        }
-      })
-    )
-    .pipe(
-      createFilterStream<SavedObject | SavedObjectsExportResultDetails>(
-        (obj) => !!obj && !(obj as SavedObjectsExportResultDetails).exportedCount
-      )
-    );
+export async function createSavedObjectsStreamFromNdJson(ndJsonStream: Readable) {
+  const savedObjects = await createPromiseFromStreams([
+    ndJsonStream,
+    createSplitStream('\n'),
+    createMapStream((str: string) => {
+      if (str && str.trim() !== '') {
+        return JSON.parse(str);
+      }
+    }),
```

Comment:
Soooo, long story short:

Uncatched errors from streams (either unhandled promises or plain errors) are causing the process to crash due to unhandled errors (forget streams were THAT much of a pain since I've been using observables)

The implementation of `createPromiseFromStreams` that is used here

https://github.com/elastic/kibana/blob/cd86b81b82f3cf28edc47f56bab900bd43214d9c/src/core/server/saved_objects/import/collect_saved_objects.ts#L49-L50

relies on `stream.pipeline` under the hood, that is supposed to intercept errors occurring during the stream/transform chain, however errors that occurred within a `pipe` of the argument observable are not.

The direct consequence is that any error thrown from the piping chain that is done in `createSavedObjectsStreamFromNdJson` was uncatched, therefor killing the server.

The correct solution would probably be to stop using streams and moving to observable instead, but these are significant changes (and `collectSavedObjects` is also used by the spaces plugin when copying to space), so the only solution I found for now was to collect the initial saved objects inside `createSavedObjectsStreamFromNdJson` and recreate a stream from that. That way, errors during the `createSavedObjectsStreamFromNdJson` stream chain are properly thrown via the added `await createPromiseFromStreams`. A FTR test has been added to assert that.

Note that this doesn't change the performances or memory usage: we were already collecting all emissions of the SO stream here before:

https://github.com/elastic/kibana/blob/cd86b81b82f3cf28edc47f56bab900bd43214d9c/src/core/server/saved_objects/import/collect_saved_objects.ts#L49

Reply:
:+1:

I suspect if we're adding more saved object types to the export (to allow us to deprecate multi-tenancy) we might want to support real streaming imports/exports again and that will maybe allow us to remove the import payload size limit completely. But that's something to revisit once we know this is a requirement.